streaming.go

```diff
@@ -25,42 +24,56 @@ import (
 )
 
 type streamingState struct {
-	protocol string
-	mu       sync.Mutex
-	attrs    []attribute.KeyValue
+	mu              sync.Mutex
+	attributes      []attribute.KeyValue
+	error           error
+	sentCounter     int
+	receivedCounter int
 }
 
 type sendReceiver interface {
 	Receive(any) error
 	Send(any) error
 }
 
-func (s *streamingState) receive(ctx context.Context, instr *instruments, msg any, conn sendReceiver) error {
+func (s *streamingState) receive(msg any, conn sendReceiver, protocol string) (int, func(), error) {
 	err := conn.Receive(msg)
 	s.mu.Lock()
-	defer s.mu.Unlock()
-	if err != nil && !errors.Is(err, io.EOF) {
-		s.attrs = append(s.attrs, statusCodeAttribute(s.protocol, err))
+	if errors.Is(err, io.EOF) {
+		return 0, s.mu.Unlock, err
 	}
+	if err != nil {
+		s.error = err
+		// If error add it to the attributes because the stream is about to terminate.
+		// If no error don't add anything because status only exists at end of stream.
+		s.attributes = append(s.attributes, statusCodeAttribute(protocol, err))
+		return 0, s.mu.Unlock, err
+	}
+	var size int
 	if msg, ok := msg.(proto.Message); ok {
-		size := proto.Size(msg)
-		instr.requestSize.Record(ctx, int64(size), s.attrs...)
+		size = proto.Size(msg)
 	}
-	instr.requestsPerRPC.Record(ctx, 1, s.attrs...)
```

Comment:
These are moved up a function call because they're dependent on the type of interceptor they're in (streaming/tracing) and what method they're in (send/receive). 

Abstracting it into here was incorrect from the beginning because of this table:

|         | Unary    | Streaming |
|---------|----------|-----------|
| Send    | Request  | Response  |
| Receive | Response | Request   |